Added Blum Goldwasser cryptosystem

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
https://en.wikipedia.org/wiki/Blum%E2%80%93Goldwasser_cryptosystem


#### Brief description of what is fixed or changed
bg_private_key, bg_public_key, encipher_bg have
been added to sympy.crypto.crypto. These methods
follow Blum–Goldwasser cryptosystem.

#### Other comments
Documentation and tests will be added soon

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* crypto
  * Added Blum Goldwasser cryptosystem
<!-- END RELEASE NOTES -->

gdp1807 committed:
Added Blum Goldwasser cryptosystem

bg_private_key, bg_public_key, encipher_bg have
been added to sympy.crypto.crypto. These methods
follow Blum–Goldwasser cryptosystem.

asmeurer commented on sympy/crypto/crypto.py:
```diff
@@ -28,7 +28,7 @@
 from sympy.utilities.misc import filldedent, translate
 from sympy.utilities.iterables import uniq
-from sympy.utilities.randtest import _randrange
+from sympy.utilities.randtest import _randrange, _randint
```
These functions are supposed to be used for the test suite. This is the same as just using `random.Random`.

czgdp1807 replied:
So should I directly use the random.Random? I saw that in GM cryptosystem _randrange was used.

czgdp1807 commented:
So should I directly use the `random.Random`? I saw that in GM cryptosystem `_randrange` was used.

gdp1807 committed:
Added methods, tests and docs for BG cryptosystem

The Blum Goldwasser cyptosystem has been added
completely. The methods, _pow_mod, bg_private_key,
bg_public_key, encipher_bg and decipher_bg has
been completed with tests and docs.

gdp1807 committed:
Update test_crypto.py

czgdp1807 commented:
@asmeurer Please review the changes I have made recently.

gdp1807 committed:
Updated test_crypto.py

gdp1807 committed:
Merge branch 'bg_crypto' of https://github.com/czgdp1807/sympy into bg_crypto

oscargus commented on sympy/crypto/crypto.py:
```diff
@@ -2246,3 +2246,199 @@ def decipher_gm(message, key):
+    Parameters
+    ==========
+
+    a: the base, a positive integer
```
Please see the examples. Should be more like
```
a : positive integer
    The base
```

oscargus commented on sympy/crypto/crypto.py:
```diff
@@ -2246,3 +2246,199 @@ def decipher_gm(message, key):
+    private keys passed as arguments and
+    then returns their product.
+
+    Parametes
```
Parameters. Missing r.

oscargus replied:
Missing an "r" in Paramete**r**s. :-)

czgdp1807 replied:
Oh yes, I missed that.

gdp1807 committed:
improved docs

gdp1807 committed:
Update crypto.py

jksuom commented on sympy/crypto/crypto.py:
```diff
@@ -2246,3 +2246,202 @@ def decipher_gm(message, key):
+
+################ Blum–Goldwasser cryptosystem  #########################
+
+def _pow_mod(a, b, m):
```
How about using Python's [builtin pow](https://docs.python.org/3.7/library/functions.html#pow)?

czgdp1807 replied:
Thanks for this. I am going to update.

gdp1807 committed:
replaced _pow_mod with pow

czgdp1807 committed:
Merged remote bg_crypto into local

czgdp1807 committed:
removed trivial error

oscarbenjamin commented:
This looks good to me. There is an unresolved comment from @asmeurer but it looks like it can be merged anyway.

I'll merge soon if no one objects.

jksuom commented on sympy/crypto/crypto.py:
```diff
@@ -2246,3 +2246,174 @@ def decipher_gm(message, key):
+        x = x**2 % key
+
+    encrypt_msg = []
+    for (m, b) in zip(enc_msg, rand_bits):
```
Would it be possible to use something like `encrypt_msg = [m^b for m, b in zip(enc_msg, rand_bits)]`?

czgdp1807 replied:
@jksuom My apologies for late reply. I think both ways are correct and I think yours is better. @oscarbenjamin If you also agree then I will make the change.

oscarbenjamin replied:
I agree with @jksuom

oscarbenjamin commented:
@czgdp1807 if you respond to @jksuom's comment I'll threaten to merge again :)

oscarbenjamin commented:
To be clear I don't know the crypto module in SymPy and I don't know the Blum Goldwasser cryptosystem. This just looks like a good patch to me and it would seem a shame for it to be abandoned. I have recently been through and labelled all 600 open PRs by topic and I saw many that looked like good patches that someone wrote that were just forgotten later. I think it's just because reviewing patches takes time and sometimes contributors are sending patches for areas that no regular contributors are focussed on.

My main concern about adopting a patch from an area I don't know that well is whether or not it represents a reasonable thing to include in SymPy at all. Three others have reviewed this patch suggesting minor improvements and no one has objected to the general concept so I will assume that this does represent a reasonable inclusion once the patch is good enough.

To me the patch looks good enough. However @jksuom has made a suggestion for improvement and it is a reasonable suggestion so if you do that and no one objects I will merge this.

czgdp1807 commented:
@oscarbenjamin Yes, I will surely update the PR. Actually, @jksuom suggestion is a good one to simplify the code. However, I thought to confirm since, any changes made will start another set of tests, which may fail due to one reason or other. Hence, confirming from two members sounded good to me. Since, we all agree I will make the changes as soon as possible.

czgdp1807 committed:
Simplified code

czgdp1807 commented:
@oscarbenjamin @jksuom I have updated the PR. Please provide reviews, if any?

oscarbenjamin commented:
Thanks for this. I'm merging it.
